### DISCUSSION TITLE
Understanding Usage

### DISCUSSION THREAD

**Discussion by AndrewF123** [startz/oauth2-etsy#2] [Q&A] [answered]
Hi, 

I've downloaded and installed the OAuth Client and Etsy Provider and have successfully got a code/state coming back in my call-back URL.  If I've understood the process is as follows:

1. Submit initial request to Etsy with appropriate detail to obtain Authorisation Code
2. Receive response back to Call-back URL with Authorisation Code and State
3. Validate State matches expected State
4. Submit Authorisation Code, Code Verifier and other appropriate details to obtain Access Token
5. Use Access Token in a request to access the API and obtain details required

From what I understand, steps 1-4 are covered by the code supplied. However, whenever I launch the code, I only get to step 2 where the call-back URL is hit with the Code and State correctly returned.  I was starting to store the details in my database to enable me to complete steps 3 and 4, however I don't know how to access the Code Verifier that is generated as part of the code in step 1.  Having revisited the code provided, I can see I should be able to access the token at the end, but can't seem to get it.  Additionally, if the code is completing an Access Token request, where can I get the Refresh Token and the Expiry?  Am I supposed to split the code up somewhere and run part of it as a second request?

Any help/guidance is greatly appreciated.

Thanks in advance.

Andrew

**Comment by AndrewF123** [comment]
Additionally, where do we set the Scopes like "listings_d" and "listings_r"?

**Reply by AndrewF123** [reply]
For anyone finding this question in future, worked this one out, add the Scopes to this section:

    // If we don't have an authorization code then get one
    $preChallenge = $provider->getPreChallenge();
    $authUrl = $provider->getAuthorizationUrl([
        'code_challenge' => $provider->getPKCE($preChallenge),
        'code_challenge_method' => 'S256',
	      'scope' => ['transactions_r']

**Comment by AndrewF123** [comment, marked as answer]
I have sorted this out now - not 100% sure if I did it right and suspect probably not, however, I have a working solution so for anyone else who is a complete novice and just needs a bit more guidance like I did, hopefully you'll find this useful if you come across this post.

1. Submit initial request to Etsy with appropriate detail to obtain Authorisation Code
2. On submission of this request, store the pre-challenge to be used at a later stage (store it in the Options table if using Wordpress)
3. Receive response back to Call-back URL with Authorisation Code and State
4. On the Call-back page, have some code to store the Authorisation Code and State (again, store in the Options table if using Wordpress)
5. Validate State matches expected State
6. Submit Authorisation Code, Code Verifier and other appropriate details to obtain Access Token - pull the Authorisation Code and pre-challenge (Code Verifier) from the Options table to put into your request.
7. Receive a response back with the Access Token, Refresh Token and Expires In values - store them in the DB (again in the Options table if using Wordpress)
8. Use Access Token in a request to access the API and obtain details required from the API
9. Complete a check to see if the Access Token has expired, if it has, use the Refresh token to obtain a new Access Token and Refresh token and update the values stored in the DB for future use.

The only part I haven't fully understood yet is if the Authorisation Code ever expires and needs refreshing.  Hopefully this will be of use to some others at some point though.

**Reply by cdburgess** [reply]
You got it! This is the correct process. It sounds like you got it all worked out correctly.
